ompi/request/req_test.c

```diff
@@ -321,30 +338,46 @@ int ompi_request_default_test_some(
     int * indices,
     ompi_status_public_t * statuses)
 {
-    size_t i, num_requests_null_inactive = 0, num_requests_done = 0;
+    size_t i, num_requests_null_inactive, num_requests_done;
     int rc = OMPI_SUCCESS;
     ompi_request_t **rptr;
     ompi_request_t *request;
-
     opal_atomic_mb();
+
+#if OPAL_ENABLE_PROGRESS_THREADS == 0
+    int do_it_once = 0;
+#endif
+
+    num_requests_null_inactive = 0;
+    num_requests_done = 0;
     rptr = requests;
     for (i = 0; i < count; i++, rptr++) {
         request = *rptr;
+#if OPAL_ENABLE_PROGRESS_THREADS == 0
+recheck_request_status:
+#endif
         if (request->req_state == OMPI_REQUEST_INACTIVE) {
             num_requests_null_inactive++;
             continue;
-        }
-        if( REQUEST_COMPLETE(request) ) {
+        } else if (REQUEST_COMPLETE(request)) {
             indices[num_requests_done++] = i;
         }
 #if OPAL_ENABLE_FT_MPI
         /* Check for dead requests due to process failure */
         /* Special case for MPI_ANY_SOURCE - Error managed below */
-        if(OPAL_UNLIKELY( ompi_request_is_failed(request) &&
-                          MPI_ERR_PROC_FAILED_PENDING == request->req_status.MPI_ERROR )) {
+        else if(OPAL_UNLIKELY( ompi_request_is_failed(request) &&
+                               MPI_ERR_PROC_FAILED_PENDING == request->req_status.MPI_ERROR )) {
             indices[num_requests_done++] = i;
         }
 #endif /* OPAL_ENABLE_FT_MPI */
+#if OPAL_ENABLE_PROGRESS_THREADS == 0
+        else if (0 == do_it_once) {
+            ++do_it_once;
+            if (0 != opal_progress()) {
+                goto recheck_request_status;
+            }
+        }
+#endif
```

Comment:
Same comment as for test_all above.

Reply:
In this case I prefer having the else branches instead of continue, it makes the code more readable (and the generated assembly should be the same). 

I have removed the use of goto in `MPI_Testsome` and `MPI_Testall`.